Zend/zend_inheritance.c

```diff
@@ -894,37 +910,28 @@ ZEND_API void zend_do_inheritance(zend_class_entry *ce, zend_class_entry *parent
 }
 /* }}} */
 
-static zend_bool do_inherit_constant_check(HashTable *child_constants_table, zval *parent_constant, zend_string *name, const zend_class_entry *iface) /* {{{ */
+static zend_bool do_inherit_constant_check(HashTable *child_constants_table, zend_class_constant_info *parent_constant, zend_string *name, const zend_class_entry *iface) /* {{{ */
 {
-	zval *old_constant;
-
-	if ((old_constant = zend_hash_find(child_constants_table, name)) != NULL) {
-		if (!Z_ISREF_P(old_constant) ||
-		    !Z_ISREF_P(parent_constant) ||
-		    Z_REFVAL_P(old_constant) != Z_REFVAL_P(parent_constant)) {
-			zend_error_noreturn(E_COMPILE_ERROR, "Cannot inherit previously-inherited or override constant %s from interface %s", ZSTR_VAL(name), ZSTR_VAL(iface->name));
-		}
+	zend_class_constant_info *old_const_info;
+
+	if ((old_const_info = zend_hash_find_ptr(child_constants_table, name)) != NULL) {
+	    if ((old_const_info != parent_constant) &&
+		(OBJ_CONST_NUM(old_const_info->ce, old_const_info->offset) !=
+		 OBJ_CONST_NUM(parent_constant->ce, parent_constant->offset))) {
+		zend_error_noreturn(E_COMPILE_ERROR, "Cannot inherit previously-inherited or override constant %s from interface %s", ZSTR_VAL(name), ZSTR_VAL(iface->name));
+	    }
```

Comment:
Also, won't private constants be inherited as well? Even if I'm wrong, a testcase like this would be nice:

```
<?php

class A {
    private const X = 1;
    public static function getX() { return B::X; }
}
class B extends A {}
A::getX();
```

This should throw an error.